x/ibc-rate-limit/ibc_middleware.go

```diff
@@ -0,0 +1,251 @@
+	authkeeper "github.com/cosmos/cosmos-sdk/x/auth/keeper"
+	bankkeeper "github.com/cosmos/cosmos-sdk/x/bank/keeper"
+	capabilitytypes "github.com/cosmos/cosmos-sdk/x/capability/types"
+	paramtypes "github.com/cosmos/cosmos-sdk/x/params/types"
+	channeltypes "github.com/cosmos/ibc-go/v3/modules/core/04-channel/types"
+	porttypes "github.com/cosmos/ibc-go/v3/modules/core/05-port/types"
+	"github.com/cosmos/ibc-go/v3/modules/core/exported"
+	"github.com/osmosis-labs/osmosis/v11/x/ibc-rate-limit/types"
+	lockupkeeper "github.com/osmosis-labs/osmosis/v11/x/lockup/keeper"
+)
+
+var (
+	_ porttypes.Middleware  = &IBCModule{}
+	_ porttypes.ICS4Wrapper = &ICS4Middleware{}
+)
+
+type ICS4Middleware struct {
+	channel       porttypes.ICS4Wrapper
+	accountKeeper *authkeeper.AccountKeeper
+	BankKeeper    *bankkeeper.BaseKeeper
+	WasmKeeper    *wasmkeeper.Keeper
+	LockupKeeper  *lockupkeeper.Keeper
+	ParamSpace    paramtypes.Subspace
+}
+
+func NewICS4Middleware(
+	channel porttypes.ICS4Wrapper,
+	accountKeeper *authkeeper.AccountKeeper, wasmKeeper *wasmkeeper.Keeper,
+	bankKeeper *bankkeeper.BaseKeeper, lockupKeeper *lockupkeeper.Keeper,
+	paramSpace paramtypes.Subspace,
+) ICS4Middleware {
+	return ICS4Middleware{
+		channel:       channel,
+		accountKeeper: accountKeeper,
+		WasmKeeper:    wasmKeeper,
+		BankKeeper:    bankKeeper,
+		LockupKeeper:  lockupKeeper,
+		ParamSpace:    paramSpace,
+	}
+}
+
+func (i *ICS4Middleware) SendPacket(ctx sdk.Context, chanCap *capabilitytypes.Capability, packet exported.PacketI) error {
+	var params types.Params
+	i.ParamSpace.GetIfExists(ctx, []byte("contract"), &params)
+	if params.ContractAddress == "" {
+		// The contract has not been configured. Continue as usual
+		return i.channel.SendPacket(ctx, chanCap, packet)
+	}
+
+	amount, denom, err := GetFundsFromPacket(packet)
+	if err != nil {
+		return sdkerrors.Wrap(err, "Rate limited SendPacket")
+	}
+	channelValue := i.CalculateChannelValue(ctx, denom)
+	err = CheckRateLimits(
+		ctx,
+		i.WasmKeeper,
+		"send_packet",
+		params.ContractAddress,
+		channelValue,
+		packet.GetSourceChannel(),
+		denom,
+		amount,
+	)
+	if err != nil {
+		return sdkerrors.Wrap(err, "Rate limited SendPacket")
+	}
+
+	return i.channel.SendPacket(ctx, chanCap, packet)
+}
+
+func (i *ICS4Middleware) WriteAcknowledgement(ctx sdk.Context, chanCap *capabilitytypes.Capability, packet exported.PacketI, ack exported.Acknowledgement) error {
+	return i.channel.WriteAcknowledgement(ctx, chanCap, packet, ack)
+}
+
+// CalculateChannelValue The value of an IBC channel. This is calculated using the denom supplied by the sender.
+//  if the denom is not correct, the transfer should fail somewhere else on the call chain
+func (i *ICS4Middleware) CalculateChannelValue(ctx sdk.Context, denom string) sdk.Int {
+	supply := i.BankKeeper.GetSupply(ctx, denom)
+	locked := i.LockupKeeper.GetModuleLockedCoins(ctx)
+	return supply.Amount.Add(locked.AmountOf(denom))
```

Comment:
Err should this be a sub? I think this is double counting the locked coin amount?

Reply:
I may be misinterpreting the result of `GetSupply`. The way I understood it is that when locking up, the funds would move from the bank module's addr to the lockup module's addr, so they would no longer be reported by `GetSupply`. 

If that's not the case, then yeah, we can either remove this or substract (depending on what we want)

Reply:
that simplifies things! I'll make the change. I'm wondering if it also makes sense to move this to the contract by doing a stargate query (not available yet, but soon!)